ember/data.go

```diff
@@ -0,0 +1,104 @@
+package ember
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"encoding/csv"
+	"fmt"
+	"strconv"
+	"strings"
+
+	gridintensity "github.com/thegreenwebfoundation/grid-intensity-go/api"
+)
+
+//go:embed co2-intensities-ember-2021.csv
+var emberData []byte
+
+type ApiClient struct {
+	data map[string]GridIntensity
+}
+
+func New() (gridintensity.Provider, error) {
+	emberData, err := getGridIntensityData()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ApiClient{data: emberData}, nil
+}
+
+func (a *ApiClient) GetAllRegionsCarbonIntensity(ctx context.Context) (map[string]float64, error) {
+	results := map[string]float64{}
+
+	for region, data := range a.data {
+		results[region] = data.EmissionsIntensityGCO2PerKWH
+	}
+
+	return results, nil
+}
+
+func (a *ApiClient) GetCarbonIntensity(ctx context.Context, region string) (float64, error) {
+	result, ok := a.data[strings.ToUpper(region)]
+	if !ok {
+		return 0, fmt.Errorf("region %q not found", region)
+	}
+
+	return result.EmissionsIntensityGCO2PerKWH, nil
+}
+
+func GetGridIntensityForCountry(countryCode string) (*GridIntensity, error) {
+	emberData, err := getGridIntensityData()
+	if err != nil {
+		return nil, err
+	}
+
+	result, ok := emberData[strings.ToUpper(countryCode)]
+	if !ok {
+		return nil, fmt.Errorf("country code %q not found", countryCode)
+	}
+
+	return &result, nil
+}
+
+func getGridIntensityData() (map[string]GridIntensity, error) {
+	data := map[string]GridIntensity{}
+
+	reader := bytes.NewReader(emberData)
+	rows, err := csv.NewReader(reader).ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, row := range rows {
+		countryCode := row[0]
+		if countryCode == "" || countryCode == "country_code" {
+			continue
+		}
+
+		year, err := strconv.Atoi(row[2])
+		if err != nil {
+			return nil, err
+		}
+
+		latestYear, err := strconv.Atoi(row[3])
+		if err != nil {
+			return nil, err
+		}
+
+		intensity, err := strconv.ParseFloat(row[4], 64)
+		if err != nil {
+			return nil, err
+		}
+
+		data[countryCode] = GridIntensity{
+			CountryCode:                  countryCode,
```

Comment:
Hey Ross - I've linked to CSV data here with both sets of country codes. If we had both sets here, we'd be able to support 2 and 3 character lookups.

https://github.com/thegreenwebfoundation/grid-intensity-exporter/issues/7#issuecomment-1163040375

The columns here are comparable, but I'm not sure if we should include the column about emissions type - there are definite differences between marginal and average carbon intensity, but that might be better managed at the provider choice instead.

Reply:
Being able to do lookups with either format would be really useful.

I'll add the 2 char codes to the current file from the CSV you provided. It would also be great if Ember can add that upstream.